File: Java/ZeroSumSubsequence.java
package com.company;

import java.util.HashSet;
import java.util.Set;

public class Main {

    public static boolean isZeroSumSubsequenceInArray(int array[]) {
        int sum = 0;
        Set<Integer> hs = new HashSet<Integer>();

        for (int i = 0; i < array.length; i++) {
            sum += array[i];
            if (sum == 0 || hs.contains(sum) || array[i] == 0) {
                return true;
            }

            hs.add(sum);
        }

        return false;
    }

    public static void main(String[] args) {
	    int arrayWith[] = {2, 8, -9, 1};
	    int arrayWithout[] = {1, 2, -9, 8};

	    System.out.println("Array with: "+isZeroSumSubsequenceInArray(arrayWith));
	    System.out.println("Array without: "+isZeroSumSubsequenceInArray(arrayWithout));
    }
}

File: NodeJS/ZeroSumSubsequence.js
function isArrayWithZeroSumSubsequence(array => {
  let set = new Set();
  set.add(0);
  let sum = 0;
  
  for(let i = 0; i < array.length; i++){
    sum = sum + array[i];
    if(set.has(sum)){
      return true;
    }

    set.add(sum);
  }
  
  return false;
});
